core/connection/openvpn.go

```diff
@@ -55,38 +54,56 @@ func ConfigureVpnClientFactory(
 			return nil, err
 		}
 
-		signer := signerFactory(consumerID)
+		signer := signerFactory(connectionParams.ConsumerID)
 
 		statsSaver := bytescount.NewSessionStatsSaver(statsKeeper)
 
 		statsSender := stats.NewRemoteStatsSender(
 			statsKeeper,
 			mysteriumAPIClient,
-			vpnSession.ID,
-			providerID,
+			connectionParams.SessionID,
+			connectionParams.ProviderID,
 			signer,
 			originalLocationCache.Get().Country,
 			time.Minute,
 		)
 
-		credentialsProvider := openvpn_session.SignatureCredentialsProvider(vpnSession.ID, signer)
+		credentialsProvider := openvpn_session.SignatureCredentialsProvider(connectionParams.SessionID, signer)
+
+		openvpnStateCallback := func(openvpnState openvpn.State) {
+			connectionState := OpenVpnStateCallbackToConnectionState(openvpnState)
+			if connectionState != nil {
+				stateChannel <- *connectionState
+			}
+
+			//this is the last state - close channel (according to best practices of go - channel writer controls channel)
+			if openvpnState == openvpn.ProcessExited {
+				close(stateChannel)
+			}
+		}
 
 		return openvpn_node.NewClient(
 			openvpnBinary,
 			vpnClientConfig,
-			state.NewMiddleware(stateCallback, statsSender.StateHandler),
+			state.NewMiddleware(openvpnStateCallback, statsSender.StateHandler),
 			openvpn_bytescount.NewMiddleware(statsSaver, 1*time.Second),
 			auth.NewMiddleware(credentialsProvider),
 		), nil
 	}
 }
 
-func channelToStateCallbackAdapter(channel chan openvpn.State) state.Callback {
-	return func(state openvpn.State) {
-		channel <- state
-		if state == openvpn.ProcessExited {
-			//this is the last state - close channel (according to best practices of go - channel writer controls channel)
-			close(channel)
-		}
+// OpenVpnStateCallbackToConnectionState maps openvpn.State to connection.State. Returns a pointer to connection.state, or nil
+func OpenVpnStateCallbackToConnectionState(input openvpn.State) *State {
```

Comment:
Should we really return a pointer here? Returning value can make the code more readable:

```go
        switch input {
        case openvpn.ConnectedState:
                return Connected
        case openvpn.ExitingState:
		return Disconnecting
	case openvpn.ReconnectingState:
		return Reconnecting
        default:
                return Unknown
	}
```
And we can introduce `Unknown` state to avoid `nil`